plugin/pkg/admission/wmftoolsenforcer/admission.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package wmftoolsenforcer
+
+import (
+	"io"
+	"strconv"
+	"time"
+
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	apierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/cache"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+func init() {
+	admission.RegisterPlugin("UidEnforcer", func(client client.Interface, config io.Reader) (admission.Interface, error) {
+		return NewUidEnforcer(client), nil
+	})
+}
+
+// plugin contains the client used by the uidenforcer admin controller
+type plugin struct {
+	*admission.Handler
+	client client.Interface
+	store  cache.Store
+}
+
+// NewSecurityContextDeny creates a new instance of the SecurityContextDeny admission controller
+func NewUidEnforcer(client client.Interface) admission.Interface {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return client.Namespaces().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Namespaces().Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&api.Namespace{},
+		store,
+		5*time.Minute,
+	)
+	reflector.Run()
+	return &plugin{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+		client:  client,
+		store:   store,
+	}
+}
+
+// This will verify the following:
+//  - Not a namespace modification from a namespaced user for their own namespace
+//  - User object has a numeric uid
+//  - Namespace object has an annotation called RunAsUser that's an integer
+//
+// If after all this there's no SecurityContext on each Container with a RunAsUser set to the same RunAsUser, it'll be set
+func (p *plugin) Admit(a admission.Attributes) (err error) {
+	if a.GetResource() != string(api.ResourcePods) && a.GetResource() != "namespace" {
+		return nil
+	}
+
+	// HACK: This should probably be a separate AdmissionController
+	hackNamespace, ok := a.GetObject().(*api.Namespace)
```

Comment:
For item 2 on pmorie's list, I was thinking you would use the ABAC policy to prevent users from modifying their namespace, not an admission controller.  Something like:
```
{"user":"bob", "resource": "pods", "namespace": "bob"}`
{"user":"bob", "resource": "services", "namespace": "bob"}`
{"user":"bob", "resource": "replicationControllers", "namespace": "bob"}`
{"user":"bob", "resource": "endpoints", "namespace": "bob"}`
{"user":"bob", "resource": "events", "namespace": "bob"}`
{"user":"bob", "resource": "secrets", "namespace": "bob"}`
{"user":"bob", "resource": "serviceAccounts", "namespace": "bob"}`
```
With a new block like that added for each new user you add.
Would that work for you, @yuvipanda ?

My reasoning is:

- use ABAC when possible rather than annotations because it is a more general mechanism and likely to be supported by project going forward better than your patch
- there are going to be other resources that you want to block users from too, which don't fit quite as neatly into this PR.
- better to whitelist allowed resources than to blacklist them.  That way, when a new feature in Kube comes out, you as the admin make an affirmative decision whether to allow users to use it.


Elaborating on that second point I just made: namespace is not the only object that is in a user's namespace which you don't want the user to have write access to.  You also would not want to user to modify `limitRange` or `resourceQuota` or `persistentVolumes`, if you start to use those concepts.